Add support for align-content on table cells.

fantasai opened the pull request:
#### c822a6a0fccc6b15d462d1aa76352ca3808e51c4
<pre>
Add support for align-content on table cells.
<a href="https://bugs.webkit.org/show_bug.cgi?id=266453">https://bugs.webkit.org/show_bug.cgi?id=266453</a>
<a href="https://rdar.apple.com/119701629">rdar://119701629</a>

Reviewed by Sammy Gill.

Updates vertical-align conditionals in table rendering
to first check align-content, and map behavior accordingly.

* LayoutTests/imported/w3c/web-platform-tests/css/css-align/blocks/align-content-table-cell-expected.txt: Update to pass
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-align/blocks/align-content-table-cell-expected.txt: Removed.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-align/blocks/align-content-table-cell-expected.txt: Removed.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutBlock): Skip alignment if we&apos;re a table cell
* Source/WebCore/rendering/RenderTableCell.cpp:
(WebCore::RenderTableCell::computeIntrinsicPadding): Map align-content values to corresponding vertical-align behavior.
(WebCore::RenderTableCell::styleDidChange): Handle align-content style changes same as vertical-align.
(WebCore::RenderTableCell::scrollbarsChanged): Handle align-content centering same as vertical-align.
* Source/WebCore/rendering/RenderTableCell.h:
(WebCore::RenderTableCell::isBaselineAligned const): Update to check align-content first.

Canonical link: <a href="https://commits.webkit.org/272373@main">https://commits.webkit.org/272373@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d3b4674910353bc66ae7a65a08ef42b39ec9603

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31425 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10096 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33122 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33922 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28477 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32195 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12444 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7345 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28104 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31764 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8505 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28059 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7324 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7480 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27965 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35264 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28569 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28418 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33617 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7555 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5583 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31460 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9217 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7381 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8246 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8066 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -572,7 +572,7 @@ void RenderBlockFlow::layoutBlock(bool relayoutChildren, LayoutUnit pageLogicalH
 
     LayoutUnit alignContentShift = 0_lu;
-    if ((!isPaginated || pageRemaining > newHeight) && (settings().alignContentOnBlocksEnabled()))
+    if ((!isPaginated || pageRemaining > newHeight) && (settings().alignContentOnBlocksEnabled()) && !isRenderTableCell())
```
If `shiftForAlignContent` should not be called on table cells, I think it would be a good idea to include a debug assert in the implementation to help guard against that in the future

fantasai replied:
There's a couple reasons I'm hesitant to do that:
- I don't think it's actually harmful atm.
- We might at some point refactor things to shift table cell alignment into here. We're not handling baseline alignment in e.g. flex and grid containers in RenderBlockFlow yet, but once we do, we'll likely want to make have table cells use the same mechanism.

So I don't think this particularly needs an assert. But a comment, as you suggest, might be helpful.

sammygill commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -572,7 +572,7 @@ void RenderBlockFlow::layoutBlock(bool relayoutChildren, LayoutUnit pageLogicalH
 
     LayoutUnit alignContentShift = 0_lu;
-    if ((!isPaginated || pageRemaining > newHeight) && (settings().alignContentOnBlocksEnabled()))
+    if ((!isPaginated || pageRemaining > newHeight) && (settings().alignContentOnBlocksEnabled()) && !isRenderTableCell())
```
I think it would also be good to have a comment that briefly explains why `shiftForAlignContent` is not able to handle table cells since it doesn't seem super obvious (to me at least)

fantasai replied:
OK

sammygill commented on Source/WebCore/rendering/RenderTableCell.cpp:
```diff
@@ -230,7 +230,20 @@ void RenderTableCell::computeIntrinsicPadding(LayoutUnit rowHeight)
+        else if (alignContent.isEndward())
+            verticalAlign = VerticalAlign::Bottom;
+    }
+    switch (verticalAlign) {
```
Since `verticalAlign` could be potentially influenced by the value of `alignContent` it would probably be nice if we could somehow reflect it in the name

fantasai replied:
OK, switched to `alignment` which is more generic.

sammygill commented on Source/WebCore/rendering/RenderTableCell.h:
```diff
@@ -273,6 +273,10 @@ inline unsigned RenderTableCell::rowIndex() const
 {
+    StyleContentAlignmentData alignContent = style().alignContent();
+    if (!alignContent.isNormal())
+        return alignContent.position() == ContentPosition::Baseline;
```
```suggestion
    if (auto alignContent = style().alignContent(); !alignContent.isNormal())
        return alignContent.position() == ContentPosition::Baseline;
```

fantasai replied:
OK

sammygill commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -572,7 +572,9 @@ void RenderBlockFlow::layoutBlock(bool relayoutChildren, LayoutUnit pageLogicalH
     LayoutUnit alignContentShift = 0_lu;
-    if ((!isPaginated || pageRemaining > newHeight) && (settings().alignContentOnBlocksEnabled()))
+    // alignment isn't supported when fragmenting
+    // table cell alignment is handled in RenderTableCell::computeIntrinsicPadding
```
WebKit style guide says start of comments should be capitalized and end should have a period

fantasai replied:
Done!

fantasai committed:
Add support for align-content on table cells.
https://bugs.webkit.org/show_bug.cgi?id=266453
rdar://119701629

Reviewed by Sammy Gill.

Updates vertical-align conditionals in table rendering
to first check align-content, and map behavior accordingly.

* LayoutTests/imported/w3c/web-platform-tests/css/css-align/blocks/align-content-table-cell-expected.txt: Update to pass
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-align/blocks/align-content-table-cell-expected.txt: Removed.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-align/blocks/align-content-table-cell-expected.txt: Removed.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutBlock): Skip alignment if we're a table cell
* Source/WebCore/rendering/RenderTableCell.cpp:
(WebCore::RenderTableCell::computeIntrinsicPadding): Map align-content values to corresponding vertical-align behavior.
(WebCore::RenderTableCell::styleDidChange): Handle align-content style changes same as vertical-align.
(WebCore::RenderTableCell::scrollbarsChanged): Handle align-content centering same as vertical-align.
* Source/WebCore/rendering/RenderTableCell.h:
(WebCore::RenderTableCell::isBaselineAligned const): Update to check align-content first.

Canonical link: https://commits.webkit.org/272373@main
